connectors.md

```diff
@@ -1,26 +1,90 @@
+
+
+The Posts model connects to both SQL and MongoDB. Title, text and authorId come from SQL, the view count comes from MongoDB.
+
+
+## What's in a connector?
+
+A connector is the piece of code that links a GraphQL server to a specific backend (eg. MySQL, MongoDB, S3, neo4j). Each backend will have its own connector. Apart from connecting the GraphQL server to a backend, connectors should also:
+
+- Batch requests together whenever it makes sense
+- Cache data fetched for the backend to avoid extra requests (at least for the duration of one query)
+- Provide a way to log information about data fetched, such as how long the request took, which things were batched together, what was fetched from the cache, how much data was fetched etc.
 
-Models in Apollo server are specific to each GraphQL object type and provide an abstraction on top of the connector. They are similar to models in ORMs. Let's say for example that you have two types, Author and Post, which are both stored in MongoDB. Rather than calling the MongoDB connector directly from your resolve funcitons, you use the Author and Post models, which use the MongoDB connector. This additional level of abstraction helps separate the data fetching logic from the GraphQL schema, which makes reusing and refactoring it easier.
 
-## How to use connectors and models in apollo server
+## What's in a model?
 
-Connectors are easy to use in apollo server, requiring just three steps:
+Models are the glue between connectors - which are backend-specific - and GraphQL types - which are app-specific. They are very similar to models in ORMs, such as Rails' Active Record.
+
+Let's say for example that you have two types, Author and Post, which are both stored in MySQL. Rather than calling the MySQL connector directly from your resolve funcitons, you should create models for Author and Post, which use the MongoDB connector. This additional level of abstraction helps separate the data fetching logic from the GraphQL schema, which makes reusing and refactoring it easier.
+
+In the example schema above, the Authors model would have the following methods:
+```
+const Author = {
+  getById(id){ ... }; // get an Author by id.
+}
+```
+
+The Posts model would have the following methods:
+```
+const Posts = {
+  getById(id){ ... }; // get Post by id
+  getByTitleContains(contains){ ... }; //get a list of posts that have a word in the title
+  getByAuthor(authorId){ ... }; // get list of posts by a certain author
+  views(postId); // get the number of views for post with ID postId (fetches from MongoDB)
+}
+```
+
+In some cases it may be a good idea for your `getById` (and other) methods to take the list of fields to be fetched as an additional argument. That way the model layer can make sure to fetch only the data required from the backend. This is especially important for types that have large fields which are not always required.
```

Comment:
@stubailo fwiw, we would probably use this. Especially for joins between tables for fields on the top type. We could have a resolver in something like `photo` but in some cases its more efficient to make the query once based on the fields needed instead of two lookups